pkg/eventpersister/persister/persister-dwh.go

```diff
@@ -0,0 +1,597 @@
+	if err != nil {
+		return nil, retriable, err
+	}
+	return evalExp, false, nil
+}
+
+func (*PersisterDwh) validateTimestamp(
+	timestamp int64,
+) error {
+	actual := time.Unix(timestamp, 0)
+	now := time.Now()
+	min := now.Add(-twentyFourHours)
+	max := now.Add(twentyFourHours)
+	if actual.Before(min) || actual.After(max) {
+		return ErrInvalidEventTimestamp
+	}
+	return nil
+}
+
+func (p *PersisterDwh) getUserEvaluation(
+	environmentNamespace,
+	userID,
+	tag,
+	featureID string,
+	featureVersion int32,
+) (*featureproto.Evaluation, error) {
+	evaluation, err := p.userEvaluationStorage.GetUserEvaluation(
+		p.ctx,
+		userID,
+		environmentNamespace,
+		tag,
+		featureID,
+		featureVersion,
+	)
+	if err != nil {
+		if err == bigtable.ErrKeyNotFound {
+			handledCounter.WithLabelValues(codeUserEvaluationNotFound).Inc()
+		} else {
+			handledCounter.WithLabelValues(codeFailedToGetUserEvaluation).Inc()
+		}
+		return nil, err
+	}
+	return evaluation, nil
+}
+
+func (p *PersisterDwh) linkGoalEventByExperiment(
+	ctx context.Context,
+	event *eventproto.GoalEvent,
+	environmentNamespace, tag string,
+) (*featureproto.Evaluation, bool, error) {
+	// List experiments with the following status RUNNING, FORCE_STOPPED, and STOPPED
+	experiments, err := p.listExperiments(ctx, environmentNamespace)
+	if err != nil {
+		return nil, true, err
+	}
+	if len(experiments) == 0 {
+		return nil, false, ErrNoExperiments
+	}
+	// Find the experiment by goal ID
+	// TODO: we must change the console UI not to allow creating
+	// multiple experiments running at the same time,
+	// using the same feature flag id and goal id
+	var experiment *exproto.Experiment
+	for _, exp := range experiments {
+		if p.findGoalID(event.GoalId, exp.GoalIds) {
+			experiment = exp
+			break
+		}
+	}
+	if experiment == nil {
+		return nil, false, ErrExperimentNotFound
+	}
+	// Get the user evaluation using the experiment info
+	ev, err := p.getUserEvaluation(
+		environmentNamespace,
+		event.UserId,
+		tag,
+		experiment.FeatureId,
+		experiment.FeatureVersion,
+	)
+	if err != nil {
+		return nil, true, err
+	}
+	return ev, false, nil
+}
+
+func (*PersisterDwh) findGoalID(id string, goalIDs []string) bool {
+	for _, goalID := range goalIDs {
+		if id == goalID {
+			return true
+		}
+	}
+	return false
+}
+
+func (p *PersisterDwh) listExperiments(
+	ctx context.Context,
+	environmentNamespace string,
+) ([]*exproto.Experiment, error) {
+	exp, err, _ := p.flightgroup.Do(
+		fmt.Sprintf("%s:%s", environmentNamespace, "listExperiments"),
+		func() (interface{}, error) {
+			experiments := []*exproto.Experiment{}
+			cursor := ""
+			for {
+				resp, err := p.experimentClient.ListExperiments(ctx, &exproto.ListExperimentsRequest{
+					PageSize:             listRequestSize,
+					Cursor:               cursor,
+					EnvironmentNamespace: environmentNamespace,
+					Statuses: []exproto.Experiment_Status{
```

Comment:
We changed how to link the goal event to the experiment. Please check the changes in the PR.
https://github.com/bucketeer-io/bucketeer/pull/159